core/src/main/scala/cats/data/NonEmptySet.scala

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright (c) 2018 Luka Jacobowitz
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cats
+package data
+
+import cats.instances.sortedSet._
+import cats.kernel._
+import cats.{Always, Eq, Eval, Foldable, Later, Now, Reducible, SemigroupK, Show}
+
+import scala.collection.immutable._
+
+final class NonEmptySet[A] private (val set: SortedSet[A]) extends AnyVal {
+
+  private implicit def ordering: Ordering[A] = set.ordering
+  private implicit def order: Order[A] = Order.fromOrdering
+
+  def +(a: A): NonEmptySet[A] = new NonEmptySet(set + a)
+  def ++(as: NonEmptySet[A]): NonEmptySet[A] = concat(as)
+
+  def concat(as: NonEmptySet[A]): NonEmptySet[A] = new NonEmptySet(set ++ as.set)
+
+  def -(a: A): SortedSet[A] = set - a
+  def map[B: Order](f: A ⇒ B): NonEmptySet[B] =
+    new NonEmptySet(SortedSet(set.map(f).to: _*)(Order[B].toOrdering))
+
+  def toNonEmptyList: NonEmptyList[A] = NonEmptyList.fromListUnsafe(set.toList)
+
+  def head: A = set.head
+  def tail: SortedSet[A] = set.tail
+  def last: A = set.last
+
+  def apply(a: A): Boolean = set(a)
+  def contains(a: A): Boolean = set.contains(a)
+
+
+  def union(as: NonEmptySet[A]) = new NonEmptySet(set.union(as.toSet))
+  def size: Int = set.size
+  def forall(p: A ⇒ Boolean): Boolean = set.forall(p)
+  def exists(f: A ⇒ Boolean): Boolean = set.exists(f)
+  def find(f: A ⇒ Boolean): Option[A] = set.find(f)
+  def collect[B](pf: PartialFunction[A, B]): Set[B] = set.collect(pf)
+  def filter(p: A ⇒ Boolean): SortedSet[A] = set.filter(p)
+  def filterNot(p: A ⇒ Boolean): SortedSet[A] = filter(t => !p(t))
```

Comment:
I think we can also add `concatMap[B](A => SortedSet[B]): SortedSet[B]` and I think it would be useful. We can also do `concatMap[B](A => NonEmptySet[B]): NonEmptySet[B]`

Reply:
Yeah, that's no doubt useful.